### PR TITLE
Add Flash Attention support on ROCM

### PR DESCRIPTION
**Pull request by xinyazhang**
This patch addresses the major limitations in our previous [PR #115981](https://github.com/pytorch/pytorch/pull/115981) through the new dedicated repository [AOTriton](https://github.com/ROCm/aotriton)

- [x] Only supports MI200 series GPU (i.e., `gcnArchName == gfx90a:sramecc+:xnack-`).
    * MI300X is supported. More architectures will be added once Triton support them.
- [x] Only supports power of two sequence lengths.
    * Now it support arbitrary sequence length
- [ ] No support for varlen APIs.
    * varlen API will be supported in future release of AOTriton
- [x] Only support head dimension 16,32,64,128.
    * Now it support arbitrary head dimension <= 256
- [x] Performance is still being optimized.
    * Kernel is selected according to autotune information from Triton.

Other improvements from AOTriton include
* Allow more flexible Tensor storage layout
* More flexible API

This is a more extensive fix to #112997

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

### CODE REVIEW

File: aten/src/ATen/native/transformers/hip/flash_attn/flash_api.hip [hunk 1]
```diff
@@ -206,102 +306,51 @@ mha_fwd(const at::Tensor &q,         // batch_size x seqlen_q x num_heads x head
...
-    }
+  at::Tensor softmax_fa_t;
+  if (return_softmax) {
+    softmax_fa_t = at::empty({batch_size, num_heads, seqlen_q, seqlen_k},
+                             at::dtype(q.dtype()).device(q.device()));
+  } else {
+    softmax_fa_t = at::empty({ 0, 0, 0, 0 }, at::dtype(q.dtype()).device(q.device()));
   }
 
-  //undo reorder tensors
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/hip/flash_attn/flash_api.hip:297]: do you no longer need to permute the output?

**Comment by drisspg** [comment]
Does this check need to be updated: https://github.com/pytorch/pytorch/blob/8d03c59d59116eb57be44477a59738566707309a/aten/src/ATen/native/transformers/cuda/sdp_utils.cpp#L186?

As well was there an existing bug that when using non multiple of 16 head dims and non power of 2 sequences there was errors on AMD for sdpa but we weren't guarding?

File: cmake/External/aotriton.cmake [hunk 2]
```diff
@@ -0,0 +1,28 @@
+if(NOT __AOTRITON_INCLUDED)
```
Comment (drisspg) [inline cmake/External/aotriton.cmake:1]: @malfet friendly ping, I imagine that doing this swap is fine but wanted to check that there isn't anything else we should aware of

File: aten/src/ATen/native/transformers/hip/flash_attn/flash_api.hip (hunk 1 above)
Comment (xinyazhang) [reply aten/src/ATen/native/transformers/hip/flash_attn/flash_api.hip:297]: Yes we don't, the only limit is the last dimension must be continuous, which is also enforced by CUTLASS implementation.

**Comment by xinyazhang** [comment]
> As well was there an existing bug that when using non multiple of 16 head dims and non power of 2 sequences there was errors on AMD for sdpa but we weren't guarding?

This is the main problem this PR is trying to address. Now the SDPA backend on AMD does not require power of 2 sequences/head dimensions anymore.

File: CMakeLists.txt [hunk 3]
```diff
@@ -736,13 +736,28 @@ if(MSVC)
   append_cxx_flag_if_supported("/utf-8" CMAKE_CXX_FLAGS)
 endif()
 
-# CAVEAT: do NOT check USE_ROCM here, because USE_ROCM is always True until
-#         include(cmake/Dependencies.cmake)
+# Note for ROCM platform:
+# 1. USE_ROCM is always ON until include(cmake/Dependencies.cmake)
+# 2. USE_CUDA will become OFF during re-configuration
+# Truth Table:
+# CUDA 1st pass: USE_CUDA=True;USE_ROCM=True, FLASH evaluetes to ON by default
```
Comment (malfet) [inline CMakeLists.txt:743]: Nit: evaluetes-> evaluates

**Review by malfet — DISMISSED** [review]
LGTM, but please fix typos in CMakeLists.txt

File: CMakeLists.txt (hunk 3 above)
Comment (xinyazhang) [reply CMakeLists.txt:743]: Sure, will wait for a complete run of CI and then fix the typos.

**Comment by drisspg** [comment]
> This is the main problem this PR is trying to address. Now the SDPA backend on AMD does not require power of 2 sequences/head dimensions anymore.

This makes sense, and do you need to update the check for mi300 support?

**Comment by xinyazhang** [comment]
> > This is the main problem this PR is trying to address. Now the SDPA backend on AMD does not require power of 2 sequences/head dimensions anymore.
> 
> This makes sense, and do you need to update the check for mi300 support?

Thank you for reminding me! I have the MI300 support check in another branch but forgot the cherrypick.

**Review by atalman — DISMISSED** [review]

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change but this needs more work to be able to land in fbcode because https://github.com/ROCm/aotriton is not available there atm.  We are working to reland this change before 2.3 release' -c weird

cc @xw285cornell @jeffdaily @jithunnair-amd @malfet

After reverting, I could import the PR right away to gather its internal signals

**Comment by jithunnair-amd** [comment]
Added to 2.3.0 milestone on @malfet's suggestion to ensure this is not left out for PyTorch2.3 release

**Review by huydhn — APPROVED** [review]
Internal build looks green now

**Comment by xw285cornell** [comment]
internal diff has landed. This PR should get autoclosed soon

**Comment by xw285cornell** [comment]
@huydhn any idea why this hasn't landed yet?

**Comment by jeffdaily** [comment]
If @huydhn is unavailable, can anyone else land this?  Does pytorchbot merge work or must it be initiated by the Meta Internal side of things?

**Comment by xw285cornell** [comment]
I'll merge it. In the worst case it'll get reverted again.

**Comment by xw285cornell** [comment]
@jeffdaily feel free to merge it. I'm not sure if you want to put a release note

**Comment by marcussacana** [comment]
This PR it was replaced by other?

**Pull request closed without merging**